src/macros.rs

```diff
@@ -126,13 +126,36 @@ macro_rules! py_run_impl {
 macro_rules! wrap_pyfunction {
     ($function:path) => {
         &|py_or_module| {
+            use $crate::derive_utils::PyFunctionArguments;
             use $function as wrapped_pyfunction;
-            $crate::impl_::pyfunction::_wrap_pyfunction(&wrapped_pyfunction::DEF, py_or_module)
+            let function_arguments: PyFunctionArguments<'_> =
+                ::std::convert::Into::into(py_or_module);
+            function_arguments.wrap_pyfunction(&wrapped_pyfunction::DEF)
         }
     };
     ($function:path, $py_or_module:expr) => {{
+        use $crate::derive_utils::PyFunctionArguments;
         use $function as wrapped_pyfunction;
-        $crate::impl_::pyfunction::_wrap_pyfunction(&wrapped_pyfunction::DEF, $py_or_module)
+        let function_arguments: PyFunctionArguments<'_> = ::std::convert::Into::into($py_or_module);
+        function_arguments.wrap_pyfunction(&wrapped_pyfunction::DEF)
+    }};
+}
+
+/// Wraps a Rust function annotated with [`#[pyfunction]`](macro@crate::pyfunction).
+///
+/// This can be used with [`PyModule::add_function`](crate::types::PyModule::add_function) to add free
+/// functions to a [`PyModule`](crate::types::PyModule) - see its documentation for more information.
+#[macro_export]
+macro_rules! wrap_pyfunction_bound {
+    ($function:path) => {
+        &|module| {
+            use $function as wrapped_pyfunction;
+            module.wrap_pyfunction(&wrapped_pyfunction::DEF)
+        }
+    };
+    ($function:path, $module:expr) => {{
+        use $function as wrapped_pyfunction;
+        $module.wrap_pyfunction(&wrapped_pyfunction::DEF)
```

Comment:
This requires the second argument (or the argument to the closure in the one-arg version) to implement `PyModuleMethods`, but there are a number of places in the docs where we pass a `Python` argument instead, such as in the [function docs](https://github.com/PyO3/pyo3/blob/main/guide/src/function.md#per-argument-options):
```rust
let f = pyo3::wrap_pyfunction!(object_length)(py).unwrap();
```
Do we still want to support this pattern?

Reply:
Yes, I think we will need to. I have an idea how this might work which might be easier for me to push as a follow-up PR.

Reply:
Actually I just pushed a final commit to this branch. Please forgive me @LilyFoote! 🙈